client/src/main/java/io/pravega/client/netty/impl/ConnectionFactoryImpl.java

```diff
@@ -114,25 +114,31 @@ public void initChannel(SocketChannel ch) throws Exception {
          });
 
         // Start the client.
-        CompletableFuture<ClientConnection> result = new CompletableFuture<>();
+        CompletableFuture<ClientConnection> connectionComplete = new CompletableFuture<>();
         try {
             b.connect(location.getEndpoint(), location.getPort()).addListener(new ChannelFutureListener() {
                 @Override
                 public void operationComplete(ChannelFuture future) {
                     if (future.isSuccess()) {
+                        //since ChannelFuture is complete future.channel() is not a blocking call.
                         Channel ch = future.channel();
                         log.debug("Connect operation completed for channel:{}, local address:{}, remote address:{}",
                                 ch.id(), ch.localAddress(), ch.remoteAddress());
-                        result.complete(handler);
+                        connectionComplete.complete(handler);
                     } else {
-                        result.completeExceptionally(new ConnectionFailedException(future.cause()));
+                        connectionComplete.completeExceptionally(new ConnectionFailedException(future.cause()));
+                        future.channel().close(); // close channel in case of failure.
                     }
                 }
             });
         } catch (Exception e) {
-            result.completeExceptionally(new ConnectionFailedException(e));
+            connectionComplete.completeExceptionally(new ConnectionFailedException(e));
         }
-        return result;
+
+        CompletableFuture<Void> channelRegisteredFuture = new CompletableFuture<>(); //check if channel is registered.
+        handler.completeWhenRegistered(channelRegisteredFuture);
```

Comment:
I think we are changing the behavior of this call unnecessarily. Previously it was returning a possibly incomplete future, and now we are trying to complete the future before returning. Why can't we return an incomplete future rather than blocking?

The problem I see is with this `completeWhenRegistered` method. It is correct that the future needs to complete when registered, but I don't see the need of blocking it until complete. Does it make sense?

Reply:
Ok, I made a second pass and I understand better now what you are trying to do. It seems fine to me.